### DISCUSSION TITLE
Packing conda environments for the Raspberry Pi

### DISCUSSION THREAD

**Discussion by stephane-caron** [upkie/upkie#434] [Tips and tricks]
If you have [configured your Upkie as a Wi-Fi access point](https://github.com/upkie/upkie/wiki/Mods#wi-fi-access-point), chances are, you have already faced the issue of installing conda packages without an Internet connection. One solution is to [connect to the Internet via a separate Wi-Fi dongle](https://github.com/orgs/upkie/discussions/114#discussioncomment-6761094), but it will become tedious if the need arises too often. In what follows, we propose a way to use the Internet connection of your host computer instead.

:warning: **Update:** [Packing pixi environments for the Raspberry Pi](https://github.com/orgs/upkie/discussions/467) works better (details below).

## With a Makefile

Here is a set of Makefile rules that:

- Download conda packages for the Raspberry Pi locally on your host computer
- Pack them into an archive
- Upload the archive to your Upkie
- SSH into your Upkie and unpack the archive to get a corresponding conda environment

### Assumptions

- `PROJECT_NAME` is set (as in the main `upkie` Makefile)
- We use [Micromamba](https://mamba.readthedocs.io/en/latest/user_guide/micromamba.html) on both sides (likely easy to port to Mamba but I haven't tested)

### Host-side rules

```makefile
clean:  ## clean up temporary files
	rm -f conda_env.tar.gz

conda_env.tar.gz:
	${MAMBA_EXE} env create -f environment.yaml -n raspios_$(PROJECT_NAME) --platform linux-aarch64 -y
	tar -zcf conda_env.tar.gz -C ${MAMBA_ROOT_PREFIX}/envs/raspios_$(PROJECT_NAME) .
	${MAMBA_EXE} env remove -n raspios_$(PROJECT_NAME) -y

.PHONY: pack_conda_env
pack_conda_env: conda_env.tar.gz  ## prepare conda environment to install it offline on your Upkie
```

### Upkie-side rule

```makefile
.PHONY: unpack_conda_env
unpack_conda_env:  ### unpack conda environment to remote conda path
	-${MAMBA_EXE} env list | grep $(PROJECT_NAME) > /dev/null && ${MAMBA_EXE} env remove -n $(PROJECT_NAME) -y
	mkdir -p ${MAMBA_ROOT_PREFIX}/envs/$(PROJECT_NAME)
	tar -zxf conda_env.tar.gz -C ${MAMBA_ROOT_PREFIX}/envs/$(PROJECT_NAME)
```

## See also

This simple method is not as general as [`conda-pack`](https://conda.github.io/conda-pack/), but `conda-pack` has the following caveat as of writing this post:

> The OS where the environment was built must match the OS of the target. This means that environments built on Windows can’t be relocated to Linux.

Unpacking an environment allows it to run on the robot, but we are only creating a virtual environment there, which does not qualify as unleashing the full power of conda :wink: (for instance we don't sync packages).

**Comment by traversaro** [comment]
> ## See also
> 
> This simple method is not as general as [`conda-pack`](https://conda.github.io/conda-pack/), but `conda-pack` has the following caveat as of writing this post:
> 
> > The OS where the environment was built must match the OS of the target. This means that environments built on Windows can’t be relocated to Linux.
> 

I just found this discussion on Google researching similar topics, in case someone is familiar/used to pixi, a possible alternative for conda-pack is pixi-pack that instead explicitly supports, cross-platform packs see https://github.com/Quantco/pixi-pack/blob/v0.2.2/README.md#cross-platform-packs .

**Reply by stephane-caron** [reply]
Thank you @traversaro :+1: We are now using `pixi-pack` [in the PPO balancer](https://github.com/upkie/ppo_balancer/blob/460c1be73e0756fe5e049e9ff857d24c748a7e79/README.md#export-dependencies-to-your-upkie) and it reduced the size of the compressed environment [from 3.3 GB to 230 MB](https://github.com/upkie/ppo_balancer/pull/5#issuecomment-2479025438) :star_struck:

Here is a separate note for further discussions: [Packing pixi environments for the Raspberry Pi](https://github.com/orgs/upkie/discussions/467).
